hadoop-hdds/docs/content/design/container-reconciliation.md

```diff
@@ -0,0 +1,344 @@
++*/
++message BlockMerkleTree {
++  optional BlockData blockData = 1; // The chunks in this should be sorted by the order of chunks written.
++  optional ChecksumData checksumData = 2; // Checksum of the checksums of the chunks.
++  optional bool deleted = 3; // If the block is deleted.
++  optional int64 length = 4; // Length of the block.
++  optional int64 chunkCount = 5; // Number of chunks in the block.
++}
++
++/*
++ContainerMerkleTree stores the checksums of the blocks in a container.
++The Container checksum is derived from the checksums of the blocks.
++Two containers across container instances on two nodes have the same checksum if they have the same set of blocks.
++If a block is deleted within the container, the checksum of the container will remain unchanged.
++ */
++message ContainerMerkleTree {
++  enum FailureCause {
++    NO_HEALTHY_CHUNK_FOUND_WITH_PEERS = 1; // No healthy chunk found with peers.
++    NO_PEER_FOUND = 2; // No peer found.
++  }
++  optional int64 containerID = 1; // The container ID.
++  repeated BlockMerkleTree blockMerkleTrees = 2; // The blocks in this should be sorted by the order of blocks written.
++  optional ChecksumData checksumData = 3; // Checksum of the checksums of the blocks.
++  optional int64 length = 5; // Length of the container.
++  optional int64 blockCount = 6; // Number of blocks in the container.
++  optional FailureCause failureCause = 7; // The cause of the failure.
++  optional int64 reconciliationCount = 8; // The reconciliation count.
++}
+
+
+```
+This is written as a file to avoid bloating the RocksDB instance which is in the IO path.
+
+## APIs
+
+The following APIs would be added to datanodes to support container reconciliation. The actions performed when calling them is defined in [Events](#events).
+
+- `reconcileContainer(containerID, List<Replica>)`
+    - Instructs a datanode to reconcile its copy of the specified container with the provided list of other container replicas.
+    - Datanodes will call `getContainerHashes` for each container replica to identify repairs needed, and use existing chunk read/write APIs to do repairs necessary.
+    - This would likely be a new command as part of the SCM heartbeat protocol, not actually a new API.
+- `getContainerHashes(containerID)`
+    - A datanode API that returns the merkle tree for a given container. The proto structure would be similar to that outlined in [Merkle Tree](#Container-Hash-Tree-/-Merkle-Tree).
+
+## Reconciliation Process
+
+SCM: Storage Container Manager
+
+DN: Datanode
+
+### SCM sets up the reconciliation process as follows:
+
+1. SCM triggers a reconciliation on `DN 1` for container 12 with replicas on `DN 2` and `DN 3`.
+   1. `SCM -> reconcileContainer(Container #12, DN2, DN3) -> DN 1` 
+
+### Datanodes set up the reconciliation process as follows:
+
+1. `DN 1` schedules the Merkle Tree to be calculated if it is not already present and report it to SCM. 
```

Comment:
Yes, we should unify these. Since we plan on calculating the merkle tree synchronously on close, the case of a container not having the tree should only be for containers created before an upgrade that the scanner has not reached yet.